src/typing.py

```diff
@@ -1128,6 +1130,8 @@ def __new__(cls, *args, **kwds):
         else:
             origin = _gorg(cls)
             obj = cls.__next_in_mro__.__new__(origin)
+            if '__dict__' in cls.__dict__:
+                obj.__orig_class__ = cls
```

Comment:
I think I'd just catch AttributeError here -- the "dict in dict" idiom feels obscure.